x-pack/plugins/elastic_assistant/server/ai_assistant_data_clients/knowledge_base/index.ts

```diff
@@ -0,0 +1,265 @@
+   *
+   * NOTE: Before automatically installing ELSER in the background, we should perform deployment resource checks
+   * Only necessary for ESS, as Serverless can always auto-install if `productTier === complete`
+   * See ml-team issue for providing 'dry run' flag to perform these checks: https://github.com/elastic/ml-team/issues/1208
+   *
+   * @param options
+   * @param options.esStore ElasticsearchStore for loading ES|QL docs via LangChain loaders
+   * @param options.soClient SavedObjectsClientContract for installing ELSER so that ML SO's are in sync
+   *
+   * @returns Promise<void>
+   */
+  public setupKnowledgeBase = async ({
+    esStore,
+    soClient,
+  }: {
+    esStore: ElasticsearchStore;
+    soClient: SavedObjectsClientContract;
+  }): Promise<void> => {
+    if (this.options.assistantService.isKBSetupInProgress) {
+      this.options.logger.debug('Knowledge Base setup already in progress');
+      return;
+    }
+
+    this.options.logger.debug('Starting Knowledge Base setup...');
+    this.options.assistantService.isKBSetupInProgress = true;
+    const elserId = await this.options.getElserId();
+
+    try {
+      const isInstalled = await this.isModelInstalled();
+      if (!isInstalled) {
+        await this.installModel({ soClient });
+        await pRetry(
+          async () =>
+            (await this.isModelInstalled())
+              ? Promise.resolve()
+              : Promise.reject(new Error('Model not installed')),
+          { minTimeout: 10000, maxTimeout: 10000, retries: 10 }
+        );
+        this.options.logger.debug(`ELSER model '${elserId}' successfully installed!`);
+      } else {
+        this.options.logger.debug(`ELSER model '${elserId}' is already installed`);
+      }
+
+      const isDeployed = await this.isModelDeployed();
+      if (!isDeployed) {
+        await this.deployModel();
+        await pRetry(
+          async () =>
+            (await this.isModelDeployed())
+              ? Promise.resolve()
+              : Promise.reject(new Error('Model not deployed')),
+          { minTimeout: 2000, retries: 10 }
+        );
+        this.options.logger.debug(`ELSER model '${elserId}' successfully deployed!`);
+      } else {
+        this.options.logger.debug(`ELSER model '${elserId}' is already deployed`);
+      }
+
+      this.options.logger.debug(`Checking if Knowledge Base docs have been loaded...`);
+      const kbDocsLoaded = (await esStore.similaritySearch(ESQL_DOCS_LOADED_QUERY)).length > 0;
+      if (!kbDocsLoaded) {
+        this.options.logger.debug(`Loading KB docs...`);
+        const loadedKnowledgeBase = await loadESQL(esStore, this.options.logger);
+        this.options.logger.debug(`${loadedKnowledgeBase}`);
+      } else {
+        this.options.logger.debug(`Knowledge Base docs already loaded!`);
+      }
+    } catch (e) {
+      this.options.logger.error(`Error setting up Knowledge Base: ${e.message}`);
```

Comment:
I simulated a failure of of the KB install and observed the error logs, but didn't observe an error toast in the browser when KB installation fails. Consider allowing errors to propagate to inform the user if they are waiting for the `Loading...` indicator to complete.

Reply:
Toast plumbed through in [00468fa](https://github.com/elastic/kibana/pull/182763/commits/00468fa5cc0537d166f69b1135185654838e3cd0)